### DISCUSSION TITLE
How to make PuLP take numbers from a predefined bucket?

### DISCUSSION THREAD

**Discussion by SanTanBan** [coin-or/pulp#499] [Q&A]
Is it possible to take numbers from a particular bucket in the case of finding integer solutions?

Say I wish to use the numbers [8,3,45,67,90,23,75,34,190]
so that the solver only uses these numbers for calculating the values of the variables which also shall be a combination of these numbers.

Presently my requirement is to use 9 integer numbers in this bucket.
I found that if these are consecutive integers, then I might use upper bound and lower bounds and create the variable as a LpInteger type. However, if the numbers are arbitrary or from a different sequence, how may I use only these numbers so that the solver them only?

**Comment by wookenny** [comment]
Usually, you can model this with additional selection variables, e.g.:
for `x` to be out of `fixed_values = [8,3,45,67,90,23,75,34,190]` you 
create binary variables:
`xs = LpVariable.dicts("xs",fixed_values, cat='Binary')`

then add constraints such that only one can be selected (set to 1) and the orginal variable `x` takes the right value
```
model += lpSum(xs) == 1
model += lpSum(value*var for value,var in xs.items()) == x
```

Now `x` has to be out of the defined set of fixed values.

**Reply by SanTanBan** [reply]
I find that this approach will not benefit me. I will not be able to decrease the number of constraints. It shall be much more beneficial if I could allow my variables to only take specific integer values without the binary approach. This is not possible now.

I am instead looking to urgently develop a new category of variables and earnestly request you to please guide me so that this can be created. The new category of variables (similar to existing "Integer", "Continuous", "Binary" may be termed as "Bucket" and the number series from which these variables will take their value may be separately initialised as an array or a generating series). This new category of variables is necessary for me to work on Discrete/Quantum Mathematics and I earnestly request you to guide me how to create this in the original PuLP module as soon as possible preferably within this year.

 I also request you to kindly enlighten me in case similar "Bucket" variables for Linear Programming are available in other Python packages or any other free softwares related to Operation Research.

**Reply by mjsaltzman** [reply]
Even if PuLP supported such a modeling idiom, the underlying solvers don't support it and don't have specialized algorithms for handling it. So all PuLP would do is convert your idiom to the formulation with binary variables to pass to the solver. They may, however, recognize the binary variable formulation and handle that with special techniques. The construction is called a _special ordered set of type 1_.

**Reply by SanTanBan** [reply]
> Even if PuLP supported such a modeling idiom, the underlying solvers don't support it and don't have specialized algorithms for handling it. So all PuLP would do is convert your idiom to the formulation with binary variables to pass to the solver. They may, however, recognize the binary variable formulation and handle that with special techniques. The construction is called a _special ordered set of type 1_.

https://github.com/coin-or/pulp/discussions/499#discussioncomment-3024386

The CoinOR CBC already has provisions for such a variable. How do I use it if not using PuLP?

**Comment by stumitchell** [comment]
I suggest you use a constraint programming approach instead of mixed integer linear programming (what pulp does)

**Reply by SanTanBan** [reply]
From what I presently understand, Constrained Optimization uses some more extra ideas (like AllDifferent, etc which may be used for a Sudoku Problem). However I am trying to know how they are able to implement it. In my case, since I was initially thinking of solving it using Operational Research softwares, I have modelled it differently.

Also I have used Python-Constraint as suggested by yourself earlier but that seemed to take much more time than PuLP-CBC. So I feel, if I can integrate my novel solution and implement it using any Exact-Solver, that would be it.

[This is because from what I consider, I would be easying the underlying implementation of AllDifferent and similar other constraints used during feasible search in CP]

However I also request you to suggest the fastest Open-Source Constraint-Solver, for trying out once more.

**Reply by SanTanBan** [reply]
I feel you are correct in terms of an immediate solution and constraint programming would solve the problem I have at hand. However, I wish to highlight the computational advantage of "No Repetition" type constraints in Constraint Programming vs using Lotsize/Bucket variables. This is because at the lower levels, I think "No Repetition" type conditions are implemented as binary variables (or maybe if-else); and this is what I am to prove the computational advantage against.

Requesting you to kindly collaborate with the Coin-OR community to develop a new type of variable which would facilitate create of the LOTSIZE variables present within the Coin-OR CBC engine. (https://github.com/coin-or/Cbc/discussions/516#discussioncomment-3821642)

**Comment by SanTanBan** [comment]
I recently posted under the CBC discussions regarding creating Bucket Variables and was answered that this already exists in COIN-OR CBC Engine. Does PuLP allow such variable definition presently?

https://github.com/coin-or/Cbc/discussions/516#discussioncomment-3007885

**Reply by SanTanBan** [reply]
Connecting with another such request on the Gurobi platform: https://support.gurobi.com/hc/en-us/community/posts/9160893511185-Requesting-a-new-type-of-variable-introduction
